File: frontend/src/import/unpack.ts
import { gunzipSync, unzipSync } from 'fflate'

/**
 * Turns an uploaded File into text. Handles .zip (first text-like entry), .gz, and plain text. All
 * in the browser — nothing is uploaded anywhere.
 */
export async function fileToText(file: File): Promise<{ text: string; innerName: string }> {
  const buf = new Uint8Array(await file.arrayBuffer())
  const name = file.name.toLowerCase()
  const isZip = name.endsWith('.zip') || (buf[0] === 0x50 && buf[1] === 0x4b)
  const isGz = name.endsWith('.gz') || (buf[0] === 0x1f && buf[1] === 0x8b)
  if (isZip) {
    const entries = unzipSync(buf)
    const candidates = Object.keys(entries).filter((n) => !n.endsWith('/') && !n.startsWith('__MACOSX'))
    const pick = candidates.sort((a, b) => entries[b].length - entries[a].length)[0]
    if (!pick) throw new Error('zip archive contains no files')
    return { text: new TextDecoder().decode(entries[pick]), innerName: pick }
  }
  if (isGz) {
    return { text: new TextDecoder().decode(gunzipSync(buf)), innerName: file.name.replace(/\.gz$/i, '') }
  }
  return { text: new TextDecoder().decode(buf), innerName: file.name }
}

export async function sha256Hex(data: Uint8Array | string): Promise<string> {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data
  const digest = await crypto.subtle.digest('SHA-256', bytes as BufferSource)
  return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, '0')).join('')
}

File: mobile/android/app/src/main/kotlin/com/hearth/genome/Unpack.kt
package com.hearth.genome

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.security.MessageDigest
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream
import java.util.zip.ZipInputStream

/**
 * A picked file as text (unpack.ts `fileToText`): a .zip gives its largest entry, a .gz its
 * contents, anything else is read as UTF-8. All on the phone; nothing is uploaded.
 */
data class Unpacked(val text: String, val innerName: String)

fun fileToText(bytes: ByteArray, fileName: String): Unpacked {
    val name = fileName.lowercase()
    val isZip = name.endsWith(".zip") || (bytes.size > 1 && bytes[0] == 0x50.toByte() && bytes[1] == 0x4b.toByte())
    val isGz = name.endsWith(".gz") || (bytes.size > 1 && bytes[0] == 0x1f.toByte() && bytes[1] == 0x8b.toByte())
    if (isZip) {
        var best: Pair<String, ByteArray>? = null
        ZipInputStream(ByteArrayInputStream(bytes)).use { z ->
            while (true) {
                val e = z.nextEntry ?: break
                if (e.isDirectory || e.name.startsWith("__MACOSX")) continue
                val data = z.readBytes()
                if (best == null || data.size > best!!.second.size) best = e.name to data
            }
        }
        val (inner, data) = best ?: error("zip archive contains no files")
        return Unpacked(decode(data), inner)
    }
    if (isGz) return Unpacked(decode(gunzip(bytes)), fileName.replace(Regex("\\.gz$", RegexOption.IGNORE_CASE), ""))
    return Unpacked(decode(bytes), fileName)
}

/** TextDecoder's defaults: UTF-8, malformed bytes replaced, a leading byte-order mark dropped. */
fun decode(bytes: ByteArray): String = String(bytes, Charsets.UTF_8).removePrefix("﻿")

fun gunzip(bytes: ByteArray): ByteArray = GZIPInputStream(ByteArrayInputStream(bytes)).use { it.readBytes() }

fun gzip(bytes: ByteArray): ByteArray = ByteArrayOutputStream().also { out -> GZIPOutputStream(out).use { it.write(bytes) } }.toByteArray()

fun sha256Hex(bytes: ByteArray): String =
    MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

fun sha256Hex(text: String): String = sha256Hex(text.toByteArray(Charsets.UTF_8))
